index.js

```diff
@@ -26,11 +26,30 @@ export default class Pbf {
 
     // === READING =================================================================
 
+    /**
+     * @template {string | number} K, V
+     * @overload
+     * @param {(tag: number, result: {key: K, value: V}, pbf: Pbf) => void} readField
+     * @param {{key: K, value: V}} result
+     * @param {number} [end]
+     * @returns {Record<K, V>}
+     */
+
+    /**
+     * @template T
+     * @overload
+     * @param {(tag: number, result: T, pbf: Pbf) => void} readField
+     * @param {T} result
+     * @param {number} [end]
+     * @returns {T}
+     */
```

Comment:
Just wondering, is this overload block necessary? It seems identical to the actual annotation below.

Reply:
It appears that `tsc` needs this repetition in JSDoc to generate function overloads.

Reply:
Hmm, weird. What if you remove the duplication but add `@overload` to the main annotation?

Reply:
TypeScript would use `any` type for `readField` and `result` arguments. This is due to TypeScript obtaining type information from the JSDoc, and since we only give it two overload annotations it would miss the implementation annotation, i.e., the TypeScript counterparts would be:

```ts
function(arg1: number);
function(arg1: string);
function(arg1: number | string) {};
```

vs

```ts
function(arg1: number);
function(arg1: string);
function(arg1) {};
```

We could have inplace type annotations instead if preferred:

```diff
diff --git a/index.js b/index.js
index 1a5d872..c06bfd4 100644
--- a/index.js
+++ b/index.js
@@ -43,15 +43,11 @@ export default class Pbf {
      * @param {number} [end]
      * @returns {T}
      */
-
-    /**
-     * @template T
-     * @param {(tag: number, result: T, pbf: Pbf) => void} readField
-     * @param {T} result
-     * @param {number} [end]
-     * @returns {T}
-     */
-    readFields(readField, result, end = this.length) {
+    readFields(
+        /** @type {(tag: number, result: any, pbf: Pbf) => void} */ readField,
+        /** @type {any} */ result,
+        end = this.length
+    ) {
         while (this.pos < end) {
             const val = this.readVarint(),
                 tag = val >> 3,
```

Reply:
Ah, let's leave it as is then, not a big deal.